exploration/bidi-usability.md

```diff
@@ -0,0 +1,329 @@
+
+>```
+> You can use {$م1صر‎ :م2صر‎ م3صر‎=م4صر‎}
+>```
+
+To prevent RTL _namespace_ names from having spillover effects with _function_ names,
+it should be possible to include "local effect" strongly directional marks in an _identifier_:
+> In this example, the _namespace_ is `:م2` and the _name_ is `:ن⁩3`, but the sequence is displayed
+> with a spillover effect.
+> (Note that the number in each name _trails_ the Arabic letter: it appears to the left because the
+> string is RTL!).
+>```
+> {$a1 :b2:c3}
+> {$م1 :م2:ن3} spillover effects
+> {⁦$م1‎ :م2‎:ن3‎⁩} with isolates and LRMs
+>```
+
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+Users cannot be expected to create or manage bidirectional controls or
+marks in _messages_, since the characters are invisible and can be difficult
+to manage.
+Tools (such as resource editors or translation editors)
+and other implementations of MessageFormat 2 serialization are strongly
+encouraged to provide paired isolates around any right-to-left
+syntax as described in this design so that _messages_ display appropriately as plain text.
+
+Ideally we do not want RLM/LRM/ALM to be part of the parsed
+`name`, `variable`, `reserved-keyword`, `unquoted`, or any other term
+defined in terms of `name`.
+This is complicated to do in ABNF because each of these tokens is followed either by
+whitespace or by some closing marker such as `}`.
+The workaround in #763 was to permit these characters _before_ or _after_ whitespace
+using the various whitespace productions.
+This works at the cost of allowing spurious markers.
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+Permit isolating bidi controls to be used on the **outside** of the following:
+- unquoted literals
+- quoted literals
+- quoted patterns
+
+This would change the ABNF as follows:
+(Notice that this change includes a production `bidi` described further down
+in this document)
+```abnf
+literal        = ( open-isolate (quoted / (unquoted [bidi])) close-isolate)
+               / (quoted / (unquoted [bidi]))
+quoted-pattern = ( open-isolate "{{" pattern "}}" close-isolate)
+               / ("{{" pattern "}}")
+
+open-isolate   = %x2066-2068
+close-isolate  = %x2069
+```
+
+> [!IMPORTANT]
+> The isolating controls go on the **_outside_** of the various _literal_ and _pattern_
+> productions because characters on the **_inside_** of these are part of the _literal_'s
+> or _pattern_'s textual content.
+> We need to allow users to include bidi controls in the output of MF2.
+
+Permit isolating bidi controls to be used **immediately inside** the following:
+- expressions
+- markup
+
+This would change the ABNF as follows (assuming the above changes are also incorporated):
+```abnf
+expression            = "{" open-isolate (literal-expression / variable-expression / annotation-expression) close-isolate "}"
```

Comment:
At the start we should only allow `%x2066`, i.e. LTR isolation.

Reply:
We permit RLI elsewhere. We *could* insist on left-to-right order. Allowing RLI might help with expressions that contain only an RTL literal:

```
{⁧لت-123-م...⁩} // RLI isolated
{لت-123-م...}
```

Reply:
While that could work for rendering, it's conceptually really rather messy. We should in that case allow the isolation of the expression contents, _and_ the literal. So we'd have:
```js
s = 'لت-123-م'
e = '{\u2066\u2067' + s + '\u2069\u2069}'
```
where the outer `\u2066..\u2069` is for the expression and the inner `\u2067..\u2069` for the literal.

My point here is that immediately within braces we're always in "code" mode, and that should _always_ be LTR. Right inside that context we could well have a RTL literal, but for that the isolation shouldn't be coming from the `expression` rule, but the `literal` one.

Reply:
I understand. I also need to add the explicit "these message things are always LTR" to the design doc.

Reply:
This is still not fixed. As expressions are code, and code is always LTR, the `expression` rule should not allow for anything other that an LRI.

Yes, this does introduce an ambiguity if the expression has a literal argument that can also be LRI-isolated, but it's pretty much the same ambiguity as we already have about whitespace.